### DISCUSSION TITLE
Timezone Offset - a missing feature!

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#1396] [Ideas]
In our appointment management app (On Time Agenda) we need to have the option to set the time zone of the Date class,
that way when our users travel from one country to another we can use the original time zone inside the app.

I know we can build a class that extend the Date and write a few IF conditions, but we would like something like this done from the core level so it's fast. We do this A LOT! and adding this on a as3 level will be quite a drawback.

I believe adding a setTimezoneOffset function would be wrong because it would not help much with daylight saving time.

We were thinking more of setting named timezone like 'Europe/London'. This should be in line with the timezone in PHP and MySQL and take care of DST changes.
setLocalTimezone(timeZoneName:String):Boolean
getLocalTimezone():String
Date.getAvailableTimezones():Vector.<String>

We should be able to get and set the timezone in this format, either:
- static, that will affect all new and existing Date() objects
or
- local on a single Date object and would override system and static time zones.
Setting null would cancel the override.

PHP handles it like this
A list of timezone name should also be provided trough a method.
We need to keep our clients in the same timezone. So we will need the static functions more.
This doesn't have to be on Date class, it could be more globally, like maybe on NativeApplication class.
This way all date related functions will be aligned and consistent. Including Date and DateTimeFormat and any other.

Proposed functions:
setLocalTimezone(timeZoneName:String):Boolean
getLocalTimezone():String
Date.getAvailableTimezones():Vector.<String>

PHP reference to all this:
https://www.php.net/manual/en/function.date-default-timezone-set.php
https://www.php.net/manual/en/timezones.europe.php
https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_convert-tz

If think we can do more to help with this, let me know

**Comment by raresn** [comment]
Hey @ajwfrost!
Do you think you can take discussions further here?
We are more than happy to help and explain or whatever you need.

**Reply by ajwfrost** [reply]
Hi @raresn  - I recall some earlier discussions about this, I have to confess to being not 100% clear on how you're proposing that this should actually work!

Currently, if you create a Date object, you have the date/time available as either the UTC values, or the "local" values which just depend on the timezone of the device you're running on i.e. using [localtime_r](https://linux.die.net/man/3/localtime_r) on most platforms (and [GetTimeZoneInformation](https://docs.microsoft.com/en-us/windows/win32/api/timezoneapi/nf-timezoneapi-gettimezoneinformation) on Windows). (See the [source code](https://github.com/adobe/avmplus/blob/master/core/Date.cpp) and implementations of `VMPI_getLocalTimeOffset` [here for Posix](https://github.com/adobe/avmplus/blob/master/VMPI/PosixPortUtils.cpp) and [here for Windows](https://github.com/adobe/avmplus/blob/master/VMPI/WinPortUtils.cpp))

I think what you're saying is that we could add a `localTimeZone` property to the `Date` class, and change the behaviour such that we no longer treat local times as being in whatever time zone the device is set to, instead we manually change the offsets to those based on the locations given? (or if reset to 'null' then we go back to using the device timezone).

One concern I have here is how we then can work out the offsets. If there are system APIs for this it may be possible but given that these are all subject to change, we run the risk of it just being wrong. For example if you set this up to be a Russian timezone then would you expect us to decide whether or not a date is GMT+2 or GMT+3 based both on the time of year and on the year itself i.e. they used to have daylight savings shifts, but now they don't? And the EU looks to be heading the same way, which means if we build something in now that has information hard-coded, it will then be wrong once each European country decides which timezone to settle on?

To avoid all those issues, it may be better to use an online service such as https://dev.timeanddate.com/time/ which should mean we can stay up to date, but would also mean that the application needs to be online in order to get the information.

At the end of the day, with the dates, I would hope that the only points you need to be dealing with a "custom" timezone are when setting or displaying the dates i.e. someone wants to add an event at 8am on 1/1/2025 in Brussels, perhaps, and later they want to know what time that event would be for someone in Shanghai..? Which means within the app itself, the UTC times could be used (like within the avmplus Date class, everything is stored and manipulated with a UTC value). So it's just about the setting/getting value which can be done with a "UTC-to-specific-timezone" sort of function... (am wondering at this point whether a couple of utility functions would be a better approach than changing the functionality of the Date class itself..) and these methods could be backed up by an online service (maybe with some fallback data which has the potential to be wrong once anything changes..)?

Would be grateful for your thoughts :-)

thanks

**Reply by raresn** [reply]
Hi, @ajwfrost . 
What you say here looks good.
In regards to the system API - we would go with that because that is what the OS reports. If that reports wrong, then the whole OS is compromise - we for one are ok with that "risk". So pulling info from the system API would be very good. 
(Online services are not cool to use when we use `new Date();` 100 times in the app)

I think it would be smart if we could have this `localTimeZone` property to be static on the `Date `class. 

An utility functions might help, but we can't see how we can get the hour without additional mathematic operations. 

So in conclusion, a `localTimeZone` static property would be exactly what we need. 

Does this help?

**Comment by Adrian-S** [comment]
I believe JAVA has this already implemented and easy to use. These examples would change the timezone to the application level.

https://www.baeldung.com/java-jvm-time-zone#3-setting-the-time-zone-from-the-application
```
TimeZone.setDefault(TimeZone.getTimeZone("Portugal"));

Calendar calendar = Calendar.getInstance();
assertEquals(calendar.getTimeZone(), TimeZone.getTimeZone("Portugal")); // should be true
```

https://stackoverflow.com/questions/2493749/how-to-set-a-jvm-timezone-properly
```
TimeZone.setDefault(TimeZone.getTimeZone("Europe/Sofia"));
```
This latest example has time zones similar to PHP time zones.

JAVA also has `setTimeZone` for `Calendar` class:
https://docs.oracle.com/javase/7/docs/api/java/util/Calendar.html#setTimeZone(java.util.TimeZone)

`Date` is missing these features since ever.

**Reply by ajwfrost** [reply]
Thanks @Adrian-S  - so looking at the implementation of `TimeZone.getOffset`:
> Returns the offset of this time zone from UTC at the specified date. If Daylight Saving Time is in effect at the specified date, the offset value is adjusted with the amount of daylight saving.
> This method returns a historically correct offset value if an underlying TimeZone implementation subclass supports historical Daylight Saving Time schedule and GMT offset changes.

So that does look like what we need, and it's supported in Android (https://developer.android.com/reference/java/util/TimeZone)

Finding a system API for Windows and Linux/Mac and iOS would also be needed: on Linux/Mac it may be possible to do this without actually affecting the device timezone (although we'd want to double-check this!): https://stackoverflow.com/questions/1223947/time-zone-conversion-c-api-on-linux-anyone

For Windows: https://docs.microsoft.com/en-us/windows/win32/api/timezoneapi/nf-timezoneapi-systemtimetotzspecificlocaltime and we may be able to get the time zone information (e.g. https://docs.microsoft.com/en-us/windows/win32/api/timezoneapi/nf-timezoneapi-enumdynamictimezoneinformation and it looks like the data is in the registry...)

And for iOS, we seem to have this capability too: https://developer.apple.com/documentation/foundation/nstimezone?language=objc
with https://developer.apple.com/documentation/foundation/nstimezone/1387242-secondsfromgmtfordate?language=objc

Okay .. so I think this should be possible then! Will just need to double-check the Linux behaviour (we can use the same on iOS and macOS it seems)...

thanks

**Reply by Adrian-S** [reply]
Sorry @ajwfrost - I don't know what is required to update inner workings of the AIR SDK. 
If this requires different programming on each OS, I understand it has multiple implications.

For iOS: https://developer.apple.com/documentation/foundation/nstimezone/1387244-defaulttimezone?language=objc "You can set it to cause the app to run as if it were in a different time zone."

For Android: https://developer.android.com/reference/java/util/TimeZone#setDefault(java.util.TimeZone) seams identical with doc from oracle, but the description is not clear for me. I also never used Java.

All the other languages have something globally (app level) to change the timezone, or to change one single `Date` object.

**Reply by raresn** [reply]
Thanks @Adrian-S  for the help here!
@ajwfrost - what do you think? Doable? :)

**Reply by raresn** [reply]
Hey @ajwfrost I know your plate is full, but i was wondering if you had a chance to think further on this task. If it is clear and if it can be put into production.
Thanks a lot!

**Reply by ajwfrost** [reply]
Sorry @raresn for not confirming: yes I think this would be viable, we'll double-check the details and then look at implementing this in our big update to the APIs.

**Reply by raresn** [reply]
@ajwfrost  - super cool! This "big update to the APIs" is there something public and planned that i could read about? Sounds very interesting.

**Reply by raresn** [reply]
@ajwfrost just a reminder.

**Reply by raresn** [reply]
@ajwfrost any news on the big update to the APIs? is there a roadmap for this? Anything i can read about it?

**Reply by ajwfrost** [reply]
It's all in progress... just a fair number of updates to the ActionScript APIs following folks' requests, and some other updates, with a corresponding change to the version number (and to the approach which may mean we start changing the numbers a bit more frequently i.e. when we're meant to given API/descriptor changes should be accompanied by version changes!)

I keep meaning to add stuff in for roadmap details! Will have to revisit that...

**Comment by ajwfrost** [comment]
@raresn it took us a lot longer in the end but we've just released a beta (50.0.0.1) that has the Timezone class in it with the functionality that I think you were after here.. are you able to check and provide any feedback, tweaks that you think may be needed etc?

thanks

**Reply by raresn** [reply]
@ajwfrost i am in vacation now and i am super pumped to see all the discussions that are happening around AIR!
Will test this ASAP when i come back to work and provide feedback!

**Thank you!**

**Reply by raresn** [reply]
@ajwfrost 
Appears to be working as it should - we will implement this as a live feature in our app in the couple of months, and we can come back with feedback/performnace review etc.

One other thing that we would need, would be the option to get the current time zone of the device (on all platoforms), without having to actually count hours and guess stuff.

Our application requires us to know the user’s original country and time zone. We need to keep his time zone intact wherever he travels to. We need to save his original time zone and enforce it on each reload. For this we might need a `TimeZone.getDefault()` or `TimeZone.getSytemTimeZone()`.

We could probably figure out a good time zone from offset alone, but it’s a bit confusing since we have multiple time zones available that define the same time zone.
(UTC+01:00) Amsterdam, Berlin, Bern, Rome, Stockholm, Vienna
 (UTC+01:00) Belgrade, Bratislava, Budapest, Ljubljana, Prague
 (UTC+01:00) Brussels, Copenhagen, Madrid, Paris
(UTC+01:00) Casablanca
(UTC+01:00) Sarajevo, Skopje, Warsaw, Zagreb
(UTC+01:00) West Central Africa
Central Europe Standard Time
Central European Standard Time

We also need to know when the user has moved out of his time zone. Since I know DST starts at different dates for different regions, we could test a few known UTC dates to determine this change, but it would be much simpler if we get the reported system time zone.

**Reply by ajwfrost** [reply]
For the system time zone, you can use a null argument to the 'getTimeZone' method:

```
                /**
                 * Obtains a TimeZone object using a string that is found in the list of <code>availableTimeZoneNames</code>.
                 * Pass in a null string to get the TimeZone object representing the current location of the device.
                 *
                 * @param timeZoneName The name of the time zone to retrieve, or null or the current device time zone.
                 * @return The TimeZone object corresponding to the requested time zone
                 *
                 * @throws ArgumentError If the requested timeZoneName is not a valid name (and isn't null)
                 *
                 * @playerversion AIR 50
                 * @langversion 3.0
                 */
                public static native function getTimeZone(timeZoneName:String):TimeZone;
```

Hmm, spotted a missing 'f' in that parameter comment.. 

To do = get this documentation generated and uploaded...!

thanks

**Reply by Adrian-S** [reply]
Works great, `TimeZone.getTimeZone(null)` resulted in the correct system time zone.

Being new to IntelliJ, we don't have any idea on how to open the parameter comment, and the documentation is on your to do list.

We did get some nasty java errors, because we forgot to change the application xml namespace to 50.0 and I don't remember having such issues ever before. Compiler did complain about that, but it took some time for us to notice. Since we only wanted a quick test, we also attempted to use Animate and Flash Builder, but with no success... as expected, since they are not supported anymore. But they were quicker to fail, and we had to restart IntelliJ a few times. Still getting used to IntelliJ.

**Reply by raresn** [reply]
Would be cool if you can tell us how to open that parameter comment thing :) 

Also, @ajwfrost not sure if you aware of this:
![image](https://user-images.githubusercontent.com/1768694/191264271-bad2c7d6-8fa9-44ff-a8d6-d1c560abf774.png)
we do get this... which we filed a bug here:
https://youtrack.jetbrains.com/issue/IDEA-275898/Argument-type-Vectorint-is-not-assignable-to-parameter-type-Vectorint
Not sure if we need to do something else here...

**Reply by ajwfrost** [reply]
@Adrian-S for the AS3 API update, you would need to change the -swf-version to 50 (and to include the airglobal.swc file) but then to run a file with swf version 50, you would need the namespace to be 50 ... but I would have hoped the error information would be clear about that..!

@raresn to get the documentation into the code hinting is also a bit of a work in progress, and it may actually vary from tool to tool sadly! Wasn't aware of that `Initializer type Vector.<String> is not assignable to variable type Vector.<String>` thing - that's very odd! wondering whether it's just code-hinting that's going wrong, or the actual compiler. Sometimes the IDEs use different tools for each... and looking at the bug you've linked to, I suspect that's what's happening here: "Despite these errors, the project compiles anyway"
